### ISSUE TITLE
liveness check for cstor-pool with respect to serving IOs

### ISSUE THREAD

**Issue by dargasudarshan** [openebs/openebs#2262]
## Description

- Application deployed on cstor(disk) went to Read-Only mode on performing a simulated node eviction(disk-pressure) post volume rebuild.

## Expected Behavior

- Volume should not go to Read-Only state when 2 pools are available.

## Current Behavior

- Volume went to read-only state.
## Steps to Reproduce

Steps Performed:
1. On a 3 node GKE cluster, deployed cstor (disk type Striped) on 2 disks attached to each node. 
2. Deployed percona mysql application with workload running along with liveness check.
```
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n chaos
NAME                       READY     STATUS    RESTARTS   AGE
percona-59d6d58b58-2hfbb   1/1       Running   0          1h
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n openebs
NAME                                                              READY     STATUS    RESTARTS   AGE
cstor-default-pool-3kum-7d8d6cdc5b-kp578                          2/2       Running   0          1h
cstor-default-pool-a18b-845f44457d-kz7nb                          2/2       Running   0          1h
cstor-default-pool-hp0q-6fc9fb57b-6mqcb                           2/2       Running   0          1h
cstor-sparse-pool-9l0s-6b9447cfc-p5hbb                            2/2       Running   0          2h
cstor-sparse-pool-s2od-5c7b4645dd-fnh4s                           2/2       Running   0          2h
cstor-sparse-pool-uw8y-65475bff6c-4rxxv                           2/2       Running   0          2h
maya-apiserver-78db8786d7-qsv2w                                   1/1       Running   0          18h
openebs-ndm-4rwm4                                                 1/1       Running   0          2h
openebs-ndm-mc77n                                                 1/1       Running   0          2h
openebs-ndm-mxn6t                                                 1/1       Running   0          2h
openebs-provisioner-6f7994f87b-wvks7                              1/1       Running   0          18h
openebs-snapshot-operator-5c998cddbd-4d6bs                        2/2       Running   0          18h
pvc-6cd21500-e3e1-11e8-9023-42010a80014d-target-86dbdcd468h87jg   3/3       Running   0          1h
```
3. After pumping ~2GB of data to the volume.
4. Detached one disk from the `cstor-default-pool-hp0q-6fc9fb57b-6mqcb ` pool.
```
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n openebs -o wide
NAME                                                              READY     STATUS    RESTARTS   AGE       IP            NODE
cstor-default-pool-3kum-7d8d6cdc5b-kp578                          2/2       Running   0          1h        10.20.7.7     gke-e2e-cstor-dev-1-default-pool-224a4771-n1tn
cstor-default-pool-a18b-845f44457d-kz7nb                          2/2       Running   0          1h        10.20.6.12    gke-e2e-cstor-dev-1-default-pool-224a4771-krsq
cstor-default-pool-hp0q-6fc9fb57b-6mqcb                           1/2       Error     0          1h        10.20.8.5     gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw
cstor-sparse-pool-9l0s-6b9447cfc-p5hbb                            2/2       Running   0          2h        10.20.6.11    gke-e2e-cstor-dev-1-default-pool-224a4771-krsq
cstor-sparse-pool-s2od-5c7b4645dd-fnh4s                           2/2       Running   0          2h        10.20.7.5     gke-e2e-cstor-dev-1-default-pool-224a4771-n1tn
cstor-sparse-pool-uw8y-65475bff6c-4rxxv                           2/2       Running   0          2h        10.20.8.4     gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw
maya-apiserver-78db8786d7-qsv2w                                   1/1       Running   0          18h       10.20.6.3     gke-e2e-cstor-dev-1-default-pool-224a4771-krsq
openebs-ndm-4rwm4                                                 1/1       Running   0          2h        10.128.0.46   gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw
openebs-ndm-mc77n                                                 1/1       Running   0          2h        10.128.0.45   gke-e2e-cstor-dev-1-default-pool-224a4771-n1tn
openebs-ndm-mxn6t                                                 1/1       Running   0          2h        10.128.0.47   gke-e2e-cstor-dev-1-default-pool-224a4771-krsq
openebs-provisioner-6f7994f87b-wvks7                              1/1       Running   0          18h       10.20.6.7     gke-e2e-cstor-dev-1-default-pool-224a4771-krsq
openebs-snapshot-operator-5c998cddbd-4d6bs                        2/2       Running   0          18h       10.20.6.10    gke-e2e-cstor-dev-1-default-pool-224a4771-krsq
pvc-6cd21500-e3e1-11e8-9023-42010a80014d-target-86dbdcd468h87jg   3/3       Running   0          1h        10.20.8.6     gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw
```
As expected pool went to degraded state
```
Use 'kubectl describe pod/cstor-default-pool-hp0q-6fc9fb57b-6mqcb -n openebs' to see all of the containers in this pod.
root@cstor-default-pool-hp0q-6fc9fb57b-6mqcb:/# zpool status
  pool: cstor-40a3d338-e3e0-11e8-9023-42010a80014d
 state: UNAVAIL
status: One or more devices are faulted in response to IO failures.
action: Make sure the affected devices are connected, then run 'zpool clear'.
   see: http://zfsonlinux.org/msg/ZFS-8000-HC
  scan: none requested
config:

        NAME                                                                                       STATE     READ WRITE CKSUM
        cstor-40a3d338-e3e0-11e8-9023-42010a80014d                                                 UNAVAIL      0     0     0  insufficient replicas
          scsi-0Google_PersistentDisk_gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw-yjl5ob-disk0  FAULTED      0     2     0  too many errors
          scsi-0Google_PersistentDisk_gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw-yjl5ob-disk1  ONLINE       0     0     0
internal error: Inappropriate ioctl for device
Aborted (core dumped)
```
5. Then re-attached the same disk back.
6. Post rebuild is completed (Observed from logs).
```
018-11-09/05:39:27.315 Rebuild started from clone for vol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d
2018-11-09/05:39:27.316 New rebuild connection
2018-11-09/05:39:27.316 Rebuild scanner started on zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d
2018-11-09/05:39:27.317 Checkpointed IO_seq: 0, Rebuild Req offset: 0, Rebuild Req length: 10737418240
2018-11-09/05:39:28.363 zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d_rebuild_clone@.io_snap0.1541741967 status change: DEGRADED -> DEGRADED
2018-11-09/05:39:28.363 zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d_rebuild_clone@.io_snap0.1541741967 rebuild status change: INIT -> INIT
2018-11-09/05:39:34.046 Checkpointed IO_seq: 0, Rebuild Req offset: 10737418240, Rebuild Req length: 10737418240
2018-11-09/05:39:37.071 Rebuilding zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d completed
2018-11-09/05:39:39.341 Checkpointed IO_seq: 0, Rebuild Req offset: 21474836480, Rebuild Req length: 10737418240
2018-11-09/05:39:44.858 Rebuilding zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d completed
2018-11-09/05:39:44.858 zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d rebuild status change: ACTIVE DATASET REBUILD INPROGRESS -> DONE
2018-11-09/05:39:44.858 zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d status change: DEGRADED -> HEALTHY
2018-11-09/05:39:44.858 closing snap cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d_rebuild_clone@.io_snap0.1541741967
2018-11-09/05:39:46.853 Rebuild process is over on zvol cstor-40a3d338-e3e0-11e8-9023-42010a80014d/pvc-6cd21500-e3e1-11e8-9023-42010a80014d
```
7. Post rebuild tainted the node on which other cstor-pool pod was running.
```
kubectl taint node gke-e2e-cstor-dev-1-default-pool-224a4771-44mn node.kubernetes.io/disk-pressure=:NoExecute
```
8. Taint caused the cstor-pool pod to evict and it remained in pending state
```
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n openebs
NAME                                                              READY     STATUS    RESTARTS   AGE
cstor-default-pool-3kum-7d8d6cdc5b-dhn2d                          0/2       Pending   0          1m
cstor-default-pool-a18b-845f44457d-kz7nb                          2/2       Running   0          1h
cstor-default-pool-hp0q-6fc9fb57b-6mqcb                           2/2       Running   1          1h
cstor-sparse-pool-9l0s-6b9447cfc-p5hbb                            2/2       Running   0          2h
cstor-sparse-pool-s2od-5c7b4645dd-rkfmz                           0/2       Pending   0          1m
cstor-sparse-pool-uw8y-65475bff6c-4rxxv                           2/2       Running   0          2h
maya-apiserver-78db8786d7-qsv2w                                   1/1       Running   0          18h
openebs-ndm-4rwm4                                                 1/1       Running   0          2h
openebs-ndm-mxn6t                                                 1/1       Running   0          2h
openebs-provisioner-6f7994f87b-wvks7                              1/1       Running   0          18h
openebs-snapshot-operator-5c998cddbd-4d6bs                        2/2       Running   0          18h
pvc-6cd21500-e3e1-11e8-9023-42010a80014d-target-86dbdcd468h87jg   3/3       Running   0          1h
```
9. At this point observed that application pod went to `CrashLoopBackOff` state, on observing the logs showed that volume went to RO state.
```
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n chaos
NAME                       READY     STATUS             RESTARTS   AGE
percona-59d6d58b58-2hfbb   0/1       CrashLoopBackOff   8          1h


find: '/var/lib/mysql/Inventory_TFwuNdqsQ0': No such file or directory
chown: cannot access '/var/lib/mysql/Inventory_TFwuNdqsQ0': No such file or directory
chown: changing ownership of '/var/lib/mysql/f{1..10}': Read-only file system
chown: changing ownership of '/var/lib/mysql/io_runner.sh': Read-only file system
chown: changing ownership of '/var/lib/mysql/s{1..10}': Read-only file system
```
## Your Environment

* `kubectl get nodes`:
```
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get nodes
NAME                                             STATUS    ROLES     AGE       VERSION
gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw   Ready     <none>    3h        v1.10.5-gke.4
gke-e2e-cstor-dev-1-default-pool-224a4771-krsq   Ready     <none>    3h        v1.10.5-gke.4
gke-e2e-cstor-dev-1-default-pool-224a4771-n1tn   Ready     <none>    3h        v1.10.5-gke.4
```
* `kubectl get pods --all-namespaces`:
```
sudarshan_darga@cloudshell:~ (strong-eon-153112)$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                                              READY     STATUS             RESTARTS   AGE
chaos         percona-59d6d58b58-2hfbb                                          0/1       CrashLoopBackOff   13         2h
kube-system   event-exporter-v0.2.1-5f5b89fcc8-zpnfr                            2/2       Running            0          19h
kube-system   fluentd-gcp-scaler-7c5db745fc-v774d                               1/1       Running            0          56m
kube-system   fluentd-gcp-v3.1.0-dxskf                                          2/2       Running            0          3h
kube-system   fluentd-gcp-v3.1.0-fmpbb                                          2/2       Running            0          3h
kube-system   fluentd-gcp-v3.1.0-nl6z4                                          2/2       Running            0          3h
kube-system   heapster-v1.5.3-5f5c74674f-txnfv                                  3/3       Running            0          56m
kube-system   kube-dns-788979dc8f-bszmk                                         4/4       Running            0          56m
kube-system   kube-dns-788979dc8f-mpjdv                                         4/4       Running            0          19h
kube-system   kube-dns-autoscaler-79b4b844b9-cljw5                              1/1       Running            0          19h
kube-system   kube-proxy-gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw         1/1       Running            0          3h
kube-system   kube-proxy-gke-e2e-cstor-dev-1-default-pool-224a4771-krsq         1/1       Running            0          3h
kube-system   kube-proxy-gke-e2e-cstor-dev-1-default-pool-224a4771-n1tn         1/1       Running            0          3h
kube-system   kubernetes-dashboard-598d75cb96-9m77f                             1/1       Running            0          19h
kube-system   l7-default-backend-5d5b9874d5-sxmx7                               1/1       Running            0          19h
kube-system   metrics-server-v0.2.1-7486f5bd67-6h76x                            2/2       Running            0          19h
openebs       cstor-default-pool-3kum-7d8d6cdc5b-dhn2d                          2/2       Running            0          56m
openebs       cstor-default-pool-a18b-845f44457d-kz7nb                          2/2       Running            0          2h
openebs       cstor-default-pool-hp0q-6fc9fb57b-6mqcb                           2/2       Running            1          2h
openebs       cstor-sparse-pool-9l0s-6b9447cfc-p5hbb                            2/2       Running            0          3h
openebs       cstor-sparse-pool-s2od-5c7b4645dd-rkfmz                           2/2       Running            0          56m
openebs       cstor-sparse-pool-uw8y-65475bff6c-4rxxv                           2/2       Running            0          3h
openebs       maya-apiserver-78db8786d7-qsv2w                                   1/1       Running            0          19h
openebs       openebs-ndm-4rwm4                                                 1/1       Running            0          3h
openebs       openebs-ndm-mxn6t                                                 1/1       Running            0          3h
openebs       openebs-ndm-np99c                                                 1/1       Running            0          37m
openebs       openebs-provisioner-6f7994f87b-wvks7                              1/1       Running            0          19h
openebs       openebs-snapshot-operator-5c998cddbd-4d6bs                        2/2       Running            0          19h
openebs       pvc-6cd21500-e3e1-11e8-9023-42010a80014d-target-86dbdcd468h87jg   3/3       Running            0          2h
```
* `kubectl get services`:
* `kubectl get sc`:
* `kubectl get pv`:
* `kubectl get pvc`:
```
kubectl get pvc --all-namespaces
NAMESPACE   NAME              STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS         AGE
chaos       demo-vol1-claim   Bound     pvc-6cd21500-e3e1-11e8-9023-42010a80014d   30G        RWO            openebs-cstor-disk   2h
```
* OS (from `/etc/os-release`):
```
sudarshan_darga@gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.5 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.5 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
* Kernel (from `uname -a`):
```
uname -a
Linux gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw 4.15.0-1015-gcp #15~16.04.1-Ubuntu SMP Thu Jul 26 20:37:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
sudarshan_darga@gke-e2e-cstor-dev-1-default-pool-224a4771-1kcw:~$
```
* Install tools:
* Others:
Openebs version : v0.7.1
Attaching the logs:

[cstor-pool-2.txt](https://github.com/openebs/openebs/files/2565059/cstor-pool-2.txt)
[cstor-pool-3.txt](https://github.com/openebs/openebs/files/2565060/cstor-pool-3.txt)
[cstor-pool-1.txt](https://github.com/openebs/openebs/files/2565061/cstor-pool-1.txt)
[target.log](https://github.com/openebs/openebs/files/2565051/target.log)
[app-ro.log](https://github.com/openebs/openebs/files/2565052/app-ro.log)
[applog.log](https://github.com/openebs/openebs/files/2565053/applog.log)

**Comment by vishnuitta** [comment]
Pool in which disk has been removed hadn't got recovered. It is stuck on some zios to get executed. One way to recover from this is - to restart the pool after disk is attached.

**Comment by ksatchit** [comment]
Updating the title based on above comments

**Comment by mittachaitu** [comment]
Performed I/O on the pool with a degraded ZFS mirror group (single disk failure in a pool) and reattached the disk, It's status shown as FAULTED in zpool status. The Latest data is not available on the reattached disk.
```
**Pool status**
zpool status 
  pool: cstor-0c478991-f920-11e8-a344-06bb7353c7fc
 state: DEGRADED
status: One or more devices are faulted in response to persistent errors.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Replace the faulted device, or use 'zpool clear' to mark the device
	repaired.
  scan: none requested
config:

	NAME                                        STATE     READ WRITE CKSUM
	cstor-0c478991-f920-11e8-a344-06bb7353c7fc  DEGRADED     0     0     0
	  mirror-0                                  DEGRADED     0     0     0
	    xvde                                    ONLINE       0     0     0
	    xvdf                                    FAULTED      0 7.18K     0  too many errors

errors: No known data errors
```
[mirrored_pool_disk_detection.txt](https://github.com/openebs/openebs/files/2652167/mirrored_pool_disk_detection.txt)

**Comment by vishnuitta** [comment]
`zpool set io.openebs:liveness_timestamp=value poolname` can be used as liveness check to verify that cstor-pool can serve IOs

**Comment by vishnuitta** [comment]
Available in 0.8.1 with this PR: https://github.com/openebs/maya/pull/894. hence, closing
